File: backend/app/domain/enums.py
"""Shared enums — single source of truth (README §5, design SSOT).

These mirror the canonical enum dictionary in design/README.md §5 and are
imported by ORM models, Pydantic schemas, services and routers alike.
"""
from __future__ import annotations

from enum import Enum


class L1Status(str, Enum):
    CREATED = "CREATED"
    FAILED = "FAILED"
    SKIPPED = "SKIPPED"
    INVALID = "INVALID"
    SUPERSEDED = "SUPERSEDED"


class L2Flag(str, Enum):
    AGREED = "agreed"
    SOFT_DISAGREEMENT = "soft_disagreement"
    HUMAN_REQUIRED = "human_required"


class L3Status(str, Enum):
    ACTIVE = "active"
    SUPERSEDED = "superseded"


class ReviewStatus(str, Enum):
    PENDING = "PENDING"
    IN_PROGRESS = "IN_PROGRESS"
    COMPLETED = "COMPLETED"
    REJECTED = "REJECTED"


class DriftStatus(str, Enum):
    NORMAL = "NORMAL"
    WARNING = "WARNING"
    CRITICAL = "CRITICAL"
    REPUBLISH_REQUIRED = "REPUBLISH_REQUIRED"


class FusionPolicy(str, Enum):
    # Canonical Label Fusion Engine default — paper Algorithm 1 (논문 §3.3, 알고리즘 1):
    #   agree(L1) → consensus(agreed); elif confidence_gap(L1) > θ → argmax_conf(soft_disagreement);
    #   else → enqueue human (L2=NULL).
    CONFIDENCE_GAP = "confidence_gap"
    MAJORITY_VOTE = "majority_vote"
    CONFIDENCE_WEIGHTED = "confidence_weighted"
    RULE_PRIORITY = "rule_priority"
    HUMAN_PRIORITY = "human_priority"
    KAPPA_BASED = "kappa_based"
    CUSTOM_POLICY = "custom_policy"


class LabelLevel(str, Enum):
    L2 = "L2"
    L3 = "L3"
    L3_PRIORITY = "L3_PRIORITY"


class LabelMethod(str, Enum):
    RULE = "rule"
    LLM = "llm"
    HUMAN = "human"


class Role(str, Enum):
    ADMIN = "Admin"
    DATA_ENGINEER = "DataEngineer"
    ML_ENGINEER = "MLEngineer"
    REVIEWER = "Reviewer"
    VIEWER = "Viewer"


# Role hierarchy: higher index = more privilege (절차 12 RBAC).
ROLE_ORDER = [
    Role.VIEWER,
    Role.REVIEWER,
    Role.ML_ENGINEER,
    Role.DATA_ENGINEER,
    Role.ADMIN,
]


def role_at_least(actual: Role, required: Role) -> bool:
    """True if ``actual`` meets or exceeds ``required`` in the privilege order.

    Reviewer and MLEngineer are siblings in the PRD matrix; we treat the linear
    order as a coarse gate and let routers add explicit role checks where the
    matrix is non-linear (e.g. Reviewer-only review completion).
    """
    return ROLE_ORDER.index(actual) >= ROLE_ORDER.index(required)

File: frontend/src/lib/enums.ts
// Shared enums mirroring backend app/domain/enums.py (design/README.md §5 SSOT).

export const ROLES = ['Viewer', 'Reviewer', 'MLEngineer', 'DataEngineer', 'Admin'] as const
export type Role = (typeof ROLES)[number]

// Privilege order (low → high), matches backend ROLE_ORDER.
export const ROLE_ORDER: Role[] = ['Viewer', 'Reviewer', 'MLEngineer', 'DataEngineer', 'Admin']

export function roleAtLeast(actual: Role | null, required: Role): boolean {
  if (!actual) return false
  return ROLE_ORDER.indexOf(actual) >= ROLE_ORDER.indexOf(required)
}

export const L1_STATUS = ['CREATED', 'FAILED', 'SKIPPED', 'INVALID', 'SUPERSEDED'] as const
export type L1Status = (typeof L1_STATUS)[number]

export const L2_FLAG = ['agreed', 'soft_disagreement', 'human_required'] as const
export type L2Flag = (typeof L2_FLAG)[number]

export const L3_STATUS = ['active', 'superseded'] as const
export type L3Status = (typeof L3_STATUS)[number]

export const REVIEW_STATUS = ['PENDING', 'IN_PROGRESS', 'COMPLETED', 'REJECTED'] as const
export type ReviewStatus = (typeof REVIEW_STATUS)[number]

export const DRIFT_STATUS = ['NORMAL', 'WARNING', 'CRITICAL', 'REPUBLISH_REQUIRED'] as const
export type DriftStatus = (typeof DRIFT_STATUS)[number]

export const FUSION_POLICY = [
  'confidence_gap', // 정본 기본값 — 논문 알고리즘 1
  'majority_vote',
  'confidence_weighted',
  'rule_priority',
  'human_priority',
  'kappa_based',
  'custom_policy',
] as const
export type FusionPolicy = (typeof FUSION_POLICY)[number]

export const LABEL_LEVEL = ['L2', 'L3', 'L3_PRIORITY'] as const
export type LabelLevel = (typeof LABEL_LEVEL)[number]

export const LABEL_METHOD = ['rule', 'llm', 'human'] as const
export type LabelMethod = (typeof LABEL_METHOD)[number]

export type StatusVariant =
  | L1Status
  | L2Flag
  | ReviewStatus
  | DriftStatus
  | L3Status
